presto-common/src/main/java/com/facebook/presto/common/block/AbstractArrayBlock.java

```diff
@@ -51,35 +55,55 @@ public String getEncodingName()
     }
 
     @Override
-    public Block copyPositions(int[] positions, int offset, int length)
+    public final Block copyPositions(int[] positions, int offset, int length)
     {
         checkArrayRange(positions, offset, length);
 
+        int oldOffsetBase = getOffsetBase();
+        int[] oldOffsets = getOffsets();
         int[] newOffsets = new int[length + 1];
-        boolean[] newValueIsNull = new boolean[length];
-
-        IntArrayList valuesPositions = new IntArrayList();
-        int newPosition = 0;
-        for (int i = offset; i < offset + length; ++i) {
-            int position = positions[i];
-            if (isNull(position)) {
-                newValueIsNull[newPosition] = true;
-                newOffsets[newPosition + 1] = newOffsets[newPosition];
+        boolean[] newValueIsNull = mayHaveNull() ? new boolean[length] : null;
+        int currentOffset = 0;
+        // First traversal will populate newValueIsNull mask (if present) and
+        // newOffsets to determine the total number of value positions involved
+        for (int i = 0; i < length; i++) {
+            newOffsets[i] = currentOffset;
+            int position = positions[i + offset];
+            if (newValueIsNull != null && isNull(position)) {
+                newValueIsNull[i] = true;
             }
             else {
-                int valuesStartOffset = getOffset(position);
-                int valuesEndOffset = getOffset(position + 1);
+                int valuesEndOffset = oldOffsets[position + oldOffsetBase + 1];
+                int valuesStartOffset = oldOffsets[position + oldOffsetBase];
                 int valuesLength = valuesEndOffset - valuesStartOffset;
+                currentOffset += valuesLength;
+            }
+        }
+        // Record last position offset
+        newOffsets[length] = currentOffset;
 
-                newOffsets[newPosition + 1] = newOffsets[newPosition] + valuesLength;
+        if (currentOffset == 0) {
+            // No elements selected
+            Block newValues = getRawElementBlock().copyRegion(getOffsetBase(), 0);
+            return createArrayBlockInternal(0, length, newValueIsNull, newOffsets, newValues);
+        }
 
-                for (int elementIndex = valuesStartOffset; elementIndex < valuesEndOffset; elementIndex++) {
-                    valuesPositions.add(elementIndex);
+        // Second traversal will use the newOffsets and total element positions to
+        // populate a correctly sized values position selection array
+        int[] elementPositions = new int[currentOffset];
+        currentOffset = 0;
+        for (int i = 0; i < length; i++) {
```

Comment:
I think this logic is harder to read than the previous version. Specifically, we would prefer a for loop over the while loop. How about just this:
```
        int currentOffset = 0;
        for (int i = 0; i < length; i++) {
            int elementsLength = newOffsets[i + 1] - newOffsets[i];
            if (elementsLength > 0) {
                int valuesStartOffset = getOffset(positions[i + offset]);
                for (int j = 0; j < elementsLength; j++) {
                    elementPositions[currentOffset++] = valuesStartOffset + j;
                }
            }
        }
```
I didn't see much gain by using direct array access to `oldOffsets` than just using `getOffset()`, and it saved 2 lines.